### PR TITLE
[PAN-OS] Add the capability to export a TSF

### PR DESCRIPTION
**Pull request by ali-sawyer**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
N/A

## Description
Add the capability to automatically export a tech support file (TSF) from PAN-OS.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:174]: why this change?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:911]: please return `CommandResults`

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:894]: is this function a general polling for the `pan-os` command?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:950]: what would happen if polling is `True` and I am trying to get an address group for example?
for that i do not need polling

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:960]: return `CommandResults`

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:3778]: default value is a command? why?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:3784]: this description is misleading a bit, because `pan-os` command can be used to retrieve any kind of data and not only jobs statuses

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:937]:
```suggestion
```

redundant

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:952]:
```suggestion
        return_results(fileResult(args.get("file_name") or "response", result.content))
```

in case that `file_name` was not provided

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:904]:
```suggestion
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:891]:
```suggestion
def pan_os_command_polling(args: dict, result: dict):
```

result is dict/list?

**Review by GuyAfik — COMMENTED** [review]
@ali-sawyer good job!

is it required to re-implement the `pan-os`? 

we want to keep it as simple as possible.

I would suggest to implement specific commands that you need with polling instead

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (MichaelYochpaz) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:289]: Let's just rename `is_pcap` here to `is_file` instead of having both, since they have the same purpose and `is_file` is more generalized.

(Make sure to change it in all the places this parameter is called, though)

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:174]: Not intentional, looks like it was a tab-to-spaces conversion by my IDE. The whitespace at the end of this string should just be removed, right?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:894]: Yes. The use case here is polling a tech support file export to keep checking until it is done. This currently has to be done using the `pan-os` command (i.e. the command below); it does not have its own command. We could add a command for it, but there are so many commands in this integration already I thought it might be best to add polling to the generic `pan-os` command.

`!pan-os type=export category=tech-support action=status job-id=<JOB_ID>`

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:950]: If the user passes in `polling=true` but no `job-id`, the command will error with "job-id is required when polling=true." because of the check at the beginning of `pan_os_command_polling()`. Do we need to handle the case where the user passes in `polling=true` and a `job-id` when polling is not needed? (Not sure how/why that would happen.)

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:3778]: Sorry did not mean to make it look like a command! The idea is just to provide a generic filename if one is not supplied. Will change it.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:3784]: Right, but the `polling` arg that this description is for only applies when polling a job status.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:960]: Will do. Please verify the change because this is from the current version of the integration.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:937]: Not sure if I'm tracking but I tried to implement this change, lmk if it is what you had in mind.

**Comment by melamedbn** [comment]
Hi @ali-sawyer,

Firstly, a big thank you for your contribution and effort on the content. I've taken some time to review the playbook you've provided, and I have a few suggestions that could help improve and streamline it:

## **Playbook: PAN-OS Export Tech Support File**:

1. **Naming Convention**: To maintain consistency with our other PAN-OS playbooks, let's update the name to `PAN-OS - Export Tech Support File` (please note the bolded hyphen).
2. **Instance Name Input**: To improve usability, consider adding a data collection task. This way, users can easily specify either 'all' or a particular instance name.
3. **Playbook Image and Documentation**: Kindly upload the playbook image under the `doc_files` folder. Don't forget to reference it in the playbook's readme for easy access.

These suggestions aim to ensure the content aligns perfectly with our standards. Should you need any clarifications or if you'd like to discuss further, please reach out.

Warm regards,

Ben

**Comment by ali-sawyer** [comment]
Thanks @melamedbn! Making some code changes that affect the playbook based on conversation with @GuyAfik. So I will update the playbook and incorporate the changes from your review as well.

**Review by melamedbn — APPROVED** [review]

**Comment by ali-sawyer** [comment]
@GuyAfik I changed `pan-os-export-tech-support-file` to be a single command as discussed on Slack.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:2]: what's this?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:842]:
```suggestion
```

this will exit the command execution and will not continue further

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:865]:
```suggestion
        'job-id': job_id,
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:876]:
```suggestion
def download_tsf(job_id: str):
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:884]:
```suggestion
        'job-id': job_id,
```

**Review by GuyAfik — COMMENTED** [review]
@ali-sawyer added a few comments, nice work! you are on the right track

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:2]: My bad, I copied the code out of the UI. Changing back.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:842]: It doesn't – I've tested it and it doesn't interrupt the flow. Let me know if I should still remove it.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:913]: I removed the download-tsf command, so there is currently no way to try to download the TSF before the job is done.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (MichaelYochpaz) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:937]: What is the purpose of this `polling` parameter? I don't see it anywhere in use inside the `export_tsf_output` besides here (which is used for passing a parameter for this function).

File: Packs/PAN-OS/pack_metadata.json
Comment (MichaelYochpaz) [inline Packs/PAN-OS/pack_metadata.json:5]: Why the double version bump? (2.1.3 --> 2.1.5)?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (MichaelYochpaz) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:161]: This parameter doesn't seem to be in-use anywhere in the added Python code?

**Review by MichaelYochpaz — COMMENTED** [review]
Hey, looks pretty good. Left a few additional notes.
Please also pull & merge from master, as the pack seems to have been updated since you originally opened the PR.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: Already fixed in latest commit.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:937]: I guess I thought it was either the convention or a requirement for a polling function. But in this case I think it makes sense to remove the argument and use polling no matter what for this command. Let me know what you think.

File: Packs/PAN-OS/pack_metadata.json
Comment (ali-sawyer) [reply Packs/PAN-OS/pack_metadata.json:5]: It got off somehow. I will update after pulling & merging from master.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:161]: This is related to updates in the integration since I opened this PR. This should get resolved after I pull & merge from master.

**Comment by MichaelYochpaz** [comment]
@ali-sawyer please note and fix the [failing validations](https://app.circleci.com/pipelines/github/demisto/content/308498/workflows/9a235256-4ea0-4cef-aa53-f58885ea2315/jobs/723603) and [mypy linting error](https://app.circleci.com/pipelines/github/demisto/content/308498/workflows/9a235256-4ea0-4cef-aa53-f58885ea2315/jobs/723602).
The mypy error is in-pat because of a mistake in my suggestion.
Let's change it to:
```python
    params = {}

    for arg in args.keys():
        params[arg] = args[arg]

    is_xml = argToBoolean(params.get("is_xml", "false"))```

Feel free to message me on Slack if you need any help.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:937]: Removing this arg so that polling happens no matter what.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:840]: remove it, its not needed

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (GuyAfik) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:172]: please revert

Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: @GuyAfik I removed the whitespace because it looks like it was a mistake in the first place – none of the other items in the `PAN_DB_URL_FILTERING_CATEGORIES` array have trailing whitespace.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: done.

**Review by MichaelYochpaz — APPROVED** [review]
Looks good to me. Well done :)
Waiting for @GuyAfik's approval for a merge.

File: Packs/PAN-OS/ReleaseNotes/2_1_11.md
Comment (GuyAfik) [inline Packs/PAN-OS/ReleaseNotes/2_1_11.md:6]:
```suggestion
- Added a new command **pan-os-export-tech-support-file** to export a tech support file (TSF) from PAN-OS.
```

File: Packs/PAN-OS/ReleaseNotes/2_1_11.md
Comment (GuyAfik) [inline Packs/PAN-OS/ReleaseNotes/2_1_11.md:7]:
```suggestion
- Fixed an issue with argument **is_xml** on command **pan-os**.
```

what was the issue? please be more specific

File: Packs/PAN-OS/Integrations/Panorama/README.md
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/README.md:9004]:
```suggestion
```

this is not relevant for the customers, only for our internal use

File: Packs/PAN-OS/Integrations/Panorama/README.md
Comment (GuyAfik) [inline Packs/PAN-OS/Integrations/Panorama/README.md:9011]: is there a context here? if not lets remove it, if yes keep it

**Review by GuyAfik — APPROVED** [review]

File: Packs/PAN-OS/Integrations/Panorama/README.md
Comment (MichaelYochpaz) [reply Packs/PAN-OS/Integrations/Panorama/README.md:9004]: @ali-sawyer please re-run the SDK's `generate-docs` command following the YAML changes (d8c73d1e59ccdd451c8e0d8edfcf1ebc43ecf6ce).

**Review by jlevypaloalto — CHANGES_REQUESTED** [review]
Hey @ali-sawyer, sorry to jump in so late, I'd like to add just a few comments before you merge.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:931]: The arg `hide_polling_outputs` should also be added to the yaml, there are many example in the repo for reference.

Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:929]: @jlevypaloalto `partial_result` actually only returns once. I will send you a screenshot of what the output looks like currently. I did remove a different line that was returning on every polling iteration. 

Re: `hide_polling_outputs`, I did not find any occurrences of this grepping through the repo. I also don't see it documented.
`grep -r "hide_polling_outputs" Packs/`

Can you point me to an example of it?

Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:929]: My bad, it's `hide_polling_output` without the `s`.

The purpose of `partial_result ` is to return on every iteration, `poll_message` does so only when `hide_polling_output` isn't in the the yaml.

Maybe `partial_result` returned once because the polling completed after one round?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:911]:
```suggestion
        job_id = dict_safe_get(result, ['response', 'result', 'job'], '')
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:929]: I see, two different `PollResult`s are returned. It’s not exactly conventional but in that case partial_result makes sense.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:923]: This will always be True?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (jlevypaloalto) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:904]:
```suggestion
        job_status = dict_safe_get(
                get_tsf_export_status(job_id),
                ['response', 'result', 'job', 'status']
        )
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py
Comment (ali-sawyer) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:923]: @jlevypaloalto Yes – in this `else` code block, if there are no errors, this will always be True.

**Review by jlevypaloalto — APPROVED** [review]

**Pull request merged**
